[SPARK-33012][BUILD][K8S] Upgrade fabric8 to 4.10.3

laflechejonathan opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This PR aims to upgrade `kubernetes-client` library to track fabric8's declared compatibility for k8s 1.18.0:
https://github.com/fabric8io/kubernetes-client#compatibility-matrix

### Why are the changes needed?
According to fabric8, 4.9.2 is incompatible with k8s 1.18.0.
  
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Not tested yet.

Commit:
Upgrade fabric8 to 4.10.3

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on resource-managers/kubernetes/core/pom.xml:
```diff
@@ -89,6 +88,12 @@
     </dependency>
     <!-- End of shaded deps. -->
 
+    <dependency>
```
Why does this become necessary, out of curiosity? looks like a downgrade. I'm just wondering if it affects anything else that depends on okhttp.

dongjoon-hyun replied:
I agree with @srowen . This looks suspicious.

laflechejonathan replied:
this was my mistake -reverted.

srowen commented on dev/deps/spark-deps-hadoop-3.2-hive-2.3:
```diff
@@ -166,14 +167,31 @@ kerby-pkix/1.0.1//kerby-pkix-1.0.1.jar
+kubernetes-model-admissionregistration/4.10.3//kubernetes-model-admissionregistration-4.10.3.jar
+kubernetes-model-apiextensions/4.10.3//kubernetes-model-apiextensions-4.10.3.jar
+kubernetes-model-apps/4.10.3//kubernetes-model-apps-4.10.3.jar
+kubernetes-model-autoscaling/4.10.3//kubernetes-model-autoscaling-4.10.3.jar
```
Are all of these additional deps needed - if they really aren't we might exclude them for tidiness.

laflechejonathan replied:
I don't think we have a choice here, it looks like kubernetes-client split their jar into smaller dependencies:
https://github.com/fabric8io/kubernetes-client/issues/2108, and since we depend on the top level `kubernetes-client`, we must pull these in.

dongjoon-hyun commented:
ok to test

dongjoon-hyun commented on resource-managers/kubernetes/core/pom.xml:
```diff
@@ -29,8 +29,7 @@
   <name>Spark Project Kubernetes</name>
   <properties>
     <sbt.project.name>kubernetes</sbt.project.name>
-    <!-- Note: Please update the kubernetes client version in kubernetes/integration-tests/pom.xml -->
```
Please do not remove this note, @laflechejonathan .

dongjoon-hyun commented:
Thank you for your first contribution, @laflechejonathan . Are you using this version in your K8s environment?

dongjoon-hyun reviewed (commented):
Thank you for your first contribution, @laflechejonathan .
1. Please don't remove the `<!-- Note: Please update the kubernetes client version in kubernetes/integration-tests/pom.xml -->`.
2. Which K8s server version are you using with this fabric8 library?
3. Although the matrix shows like that, Apache Spark `master` branch seems to work with K8s v1.18.8 without this upgrade, doesn't it?

SparkQA commented:
**[Test build #129191 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129191/testReport)** for PR 29888 at commit [`ee845e0`](https://github.com/apache/spark/commit/ee845e0c7926c628bd39f1a597840436824ecac2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33806/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33806/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33806/
Test FAILed.

SparkQA commented:
**[Test build #129191 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129191/testReport)** for PR 29888 at commit [`ee845e0`](https://github.com/apache/spark/commit/ee845e0c7926c628bd39f1a597840436824ecac2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129191/
Test PASSed.

dongjoon-hyun reviewed (commented):
The K8s IT failure looks relevant a little.
```
KubernetesSuite:
...
- Run in client mode. *** FAILED ***
  io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://192.168.39.171:8443/api/v1/namespaces/2899456ab5484196859800707ad72210/pods. Message: No API token found for service account "default", retry after the token is automatically created and added to the service account. Received status: Status(apiVersion=v1, code=500, details=StatusDetails(causes=[], group=null, kind=serviceaccounts, name=create pod, retryAfterSeconds=1, uid=null, additionalProperties={}), kind=Status, message=No API token found for service account "default", retry after the token is automatically created and added to the service account, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=ServerTimeout, status=Failure, additionalProperties={}).
  at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:589)
  at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:528)
  at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:492)
  at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:451)
  at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:252)
  at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:844)
  at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:352)
  at io.fabric8.kubernetes.client.dsl.base.BaseOperation.lambda$createNew$0(BaseOperation.java:365)
  at io.fabric8.kubernetes.api.model.DoneablePod.done(DoneablePod.java:26)
  at 
```

Commit:
remove okhttp pin, regenerate manifests

SparkQA commented:
**[Test build #129269 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129269/testReport)** for PR 29888 at commit [`9aa0bdb`](https://github.com/apache/spark/commit/9aa0bdb8b77de9087baf4934e856d364a6d14b68).

laflechejonathan commented:
I believe I've addressed your comments, thank you for the quick review!

@dongjoon-hyun, to answer your questions:

1. Done!
2. We are running against k8s 1.18.8
3. Correct, in spite of the declared compatibility, spark appears to run fine against k8s 1.18.x. That said, in our other java projects that rely on `kubernetes-client`, we routinely encounter subtle breaks and behaviour changes as we upgrade, so would like to keep spark as up to date as possible, so when break does occur, the version delta is smaller and less risky.

On the integration test failure, I'm fairly certain it's a flake. We used to run into this error when launching spark applications in production, and had to add functionality to block spark-submit until the service account token is created for a namespace. Would be good to add similar functionality to solve this flake, but should be out of scope of this PR.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33886/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33886/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33886/
Test PASSed.

SparkQA commented:
**[Test build #129269 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129269/testReport)** for PR 29888 at commit [`9aa0bdb`](https://github.com/apache/spark/commit/9aa0bdb8b77de9087baf4934e856d364a6d14b68).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129269/
Test PASSed.

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @laflechejonathan and @srowen .
Merged to master for Apache Spark 3.1.0 on December 2020.

dongjoon-hyun commented:
Thank you again for your first contribution, @laflechejonathan .
I added you to the Apache Spark contributor group and assigned you to SPARK-33012.
Welcome to the Apache Spark community.
